Updated Decaf Guide (#182)
pulsar-edit/.github: guides/how-to-decaf.md

This guide should teach you the following:

  * Know what Decaffeination is.
  * Best Practices to decaf code.
  * Best Practices to decaf code for Pulsar.
  * How to make your decaf work easy to review for maintainers.

## Overview

This section covers the general overview and workflow you should adopt to successfully decaf some code.

But first to do so, there is one big rule:

  * No matter what, any PR that decafs **both** the source files of a repo/package **and** the specs cannot be accepted. Since decaffing code can be tricky, we need to ensure we catch any errors that are being produced. If both the specs are decaffed at the same time the source files are, it is impossible to tell if the process happened successfully, and which aspect is producing any test failures or errors we see later on. Generally, if a repo/package needs both the specs and source files decaffed, they should each occur as their own unique PRs.

Secondly, here is a general recommendation when choosing what to decaf:

>   * When you decide to decaf a repo/package, it's a good idea to decaf the entirety of what you are working on. Decaffeinating only a single spec file or single source file doesn't do us much good, and in many cases may be totally invalid on it's own, so it's highly recommended to, within a single PR, decaf the entire source files, or entire spec files.

Reviewer: ```suggestion
  * When you decide to decaf a repo/package, it's a good idea to decaf the entirety of what you are working on. Decaffeinating only a single spec file or single source file doesn't do us much good, and in many cases may be totally invalid on its own, so it's highly recommended to, within a single PR, decaf the entirety of the source or spec files.
```

> With those rules/recommendations out of the way, following below is the general guide that should be followed:

Reviewer: ```suggestion
With those rules/recommendations out of the way, below is the general guide that should be followed:
```

> 1. As you might expect, decide what you are going to decaffeinate. To help make everyone's lives easier, you should ensure the file you are decaffing doesn't have any open PRs against it, or be at risk of receiving major changes anytime soon. If you aren't sure of this, feel free to ask the Pulsar maintainers.

Reviewer: ```suggestion
1. As you might expect, decide what you are going to decaffeinate. To help make everyone's lives easier, you should ensure the file you are decaffing doesn't have any open PRs against it, or that it is at risk of receiving major changes anytime soon. If you aren't sure of this, feel free to ask the Pulsar maintainers.
```

2. Once you've decided what repo/package you are going to decaf, then decide if you want to decaf the specs or source files.

> 3. Now with the choices out of the way, you can get started. You'll want to make sure to create the Machine Decaf branch (more on that later) but this would be a branch that exists on the remote repo that contains only machine decaffed files. (If you don't have permissions to make this, feel free to ask the Pulsar maintainers). While we could be done there, the machine decaffed code is valid, but very unreadable and difficult to work with, so it requires manual intervention to improve.

Reviewer: This seems confusing to me, are we saying that the machine decaf branch exists on the original repo before one can start? This seems somewhat difficult for people to get involved with.

Author: While I know that's a hard barrier to entry it does. 
The remote repository needs to have a machine decaf branch first, so that the contributor doing the manual decaf work is able to create a branch from those changes (so they have the machine decaf), and more importantly, it's the branch they target to merge their manual decaf. 

Since, there's no way to target a branch that doesn't exist on remote in a PR that's why it needs to already exist. Since if we don't have a machine decaf and manual decaf diff, these are impossible to review.

That's why I mentioned that if the contributor doesn't have permissions that they can contact a Pulsar maintainer to have that branch made, as this is how I've been doing all my decaf work, and how it's being done for PPM. Before the latest PPM decaf PRs were made I had created a machine decaf branch for the contributor to target.

Reviewer: Ok I get you, I was thinking that an alternative is to submit a PR to create a machine decaf branch as a separate PR from the manual work but thinking back I realise that isn't possible without having access first as you can only compare branches in a PR, not target the creation of a new one.

Author: Yeah, I wish there was a better way to do this.

We could possibly instruct to create the machine decaf as a single commit, then only commits afterwards can add manual decaf steps. And we review a PR based on the diff between commits, but that sounds like it's much more likely to be done improperly, but even so not sure that's worth the added ability of anyone being able to work on it, without having to wait on us.

Reviewer: I agree, I think what you suggest is probably best.

4. To start you manual decaffeination, create a fork of the machine decaf files branch, and apply any changes needed on top of it (more on that later). Once you've applied all changes you need to, to improve the machine decaf work you can finally submit a PR.

5. When submitting the PR of your manual decaf work, ensure the target branch is the machine decaf branch. Having this be the target allows maintainers to quickly and painlessly review your work, and check for any actual changes that you've made.

6. Finally, once your manual decaf work is approved, it should be merged directly into the `master`/`main` branch on that repo. So that your decaf changes are applied directly to the source. From here you can cleanup by deleting the machine decaf branch.

## Machine Decaf

Like mentioned earlier, using automated tools to decaf CoffeeScript produces valid, yet unreadable and unmaintainable code.

To produce this new code in a way that we know is functional, you'll want to keep lookout for a few things:

> First note, that in general using CoffeeScript 2.x (the latest version as of this writing) is totally valid. But if you find any tests are failing with only machine decaf code, it may be because there are very few edge cases where CoffeeScript code decafs differently from 2.x tooling and 1.x tooling. Which because nearly all the CoffeeScript for Atom was written back in the CoffeeScript 1.x days, we need to account for this. You can always check if a particular repo/package is using 1.x CoffeeScript tooling by checking the version of `coffee-script` or `coffeescript` within the relevant `package.json` either `dependencies` or `devDependencies`. If those are within the 1.x range, you should use older tooling to decaf. While keeping in mind that the 1.x CoffeeScript decaf tooling will usually strip any comments from the code, and will have to be added back in.

Reviewer: ```suggestion
First note, that in general using CoffeeScript 2.x (the latest version as of this writing) is totally valid. But if you find any tests are failing with only machine decaf code, it may be because there are very few edge cases where CoffeeScript code decafs differently from 2.x tooling and 1.x tooling. Which, because nearly all the CoffeeScript for Atom was written back in the CoffeeScript 1.x days, we need to account for. You can always check if a particular repo/package is using 1.x CoffeeScript tooling by checking the version of `coffee-script` or `coffeescript` within the relevant `package.json` either `dependencies` or `devDependencies`. If those are within the 1.x range, you should use older tooling to decaf. While keeping in mind that the 1.x CoffeeScript decaf tooling will usually strip any comments from the code, and will have to be added back in.
```
